src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+namespace TorchSharp
+{
+    /// <summary>
+    /// Manages dispose scopes, that can make automatic tensor disposal easier. Note that the
+    /// DisposeManager is thread local. The DisposeScopeManager can also manage other disposables, such as training
+    /// batches and the like.
+    /// </summary>
+    public class DisposeScopeManager
+    {
+        private int _createdCount;
+        private int _disposedCount;
+
+        static DisposeScopeManager()
+        {
+            torch.Tensor.OnTensorCreateCallback.Add(TryRegisterInDisposeScope);
+            torch.Tensor.OnTensorDisposeCallback.Add(TryUnregisterInDisposeScope);
+        }
+
+        [ThreadStatic] private static DisposeScopeManager _singleton;
+
+        public static DisposeScopeManager Singleton = (_singleton ??= new DisposeScopeManager());
+
+        internal Stack<DisposeScope> DisposeScopeStack { get; } = new();
+        internal IDisposable CurrentlyDisposing { get; set; }
+
+        public int CreatedCount => _createdCount;
+        public int DisposedCount => _disposedCount;
+        public int TotalCount => _createdCount - _disposedCount;
+
+        public static long TotalAllocatedSize => Singleton.InnerTotalAllocatedSize;
+
+        internal static void TryUnregisterInDisposeScope(IDisposable disposable) =>
+            Singleton.InnerTryUnregisterInDisposeScope(disposable);
+
+        internal static void TryRegisterInDisposeScope(IDisposable disposable) =>
+            Singleton.InnerTryRegisterInDisposeScope(disposable);
+
+        private long InnerTotalAllocatedSize {
+            get {
+                long size = 0;
+                lock (this) {
```

Comment:
Why do you need to lock if all instances are thread-local?

Reply:
If it has to do with finalizers, something needs to be reworked, as finalizers should never call into neither `DisposeScope` nor `DisposeScopeManager`. Perhaps `Tensor` does not implement [the dispose pattern](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#dispose-and-disposebool) correctly.

Reply:
We had  an issue where we suspected the GC thread of doing stuff, but I think that was a mistake. If we're safe to assume that no external thread will every dispose of any of our tensors, the locking is not needed. I'll remove it for now.